src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+// guidelines that we agreed upon:
+//
+// 1. events should be nested like {"type": "string", "value": {...}} since
+// that simplifies their processing, especially in golang
+//
+// 2. the event key should be such that one can "switch" using the key and no
+// other comparison in theory should be needed to understand the event type
+//
+// Currently emitted events do not follow the above guidelines because we've
+// not finished specifying the events format.
+
+static nlohmann::json make_log_event(uint32_t verbosity, const char *message) {
+    auto verbosity_tuple = verbosity_itoa(verbosity);
+    assert(std::get<1>(verbosity_tuple));
+    const std::string &vs = std::get<0>(verbosity_tuple);
+    return nlohmann::json{
+            {"type", "LOG"}, {"verbosity", vs}, {"message", message}};
+}
+
+static nlohmann::json make_failure_event(const Error &error) {
+    return nlohmann::json{{"type", "FAILURE"}, {"failure", error.reason}};
+}
+
+static bool is_event_valid(const std::string &str) {
+    bool rv = false;
+    do {
+#define CHECK(value)                                                           \
+    if (#value == str) {                                                       \
+        rv = true;                                                             \
+        break;                                                                 \
+    }
+        MK_ENUM_EVENT(CHECK)
+#undef CHECK
+    } while (0);
+    return rv;
+}
+
+static nlohmann::json known_events() {
+    nlohmann::json json;
+#define ADD(name) json.push_back(#name);
+    MK_ENUM_EVENT(ADD)
+#undef ADD
+    return json;
+}
+
+static std::string known_tasks() {
+    nlohmann::json json;
+#define ADD(name) json.push_back(#name);
+    MK_ENUM_TASK(ADD)
+#undef ADD
+    return json.dump();
+}
+
+static std::string known_verbosity_levels() {
+    nlohmann::json json;
+#define ADD(name) json.push_back(#name);
+    MK_ENUM_VERBOSITY(ADD)
+#undef ADD
+    return json.dump();
+}
+
+static std::unique_ptr<nettests::Runnable> make_runnable(const std::string &s) {
+    std::unique_ptr<nettests::Runnable> runnable;
+#define ATOP(value)                                                            \
+    if (s == #value) {                                                         \
+        runnable.reset(new nettests::value##Runnable);                         \
+    }
+    MK_ENUM_TASK(ATOP)
+#undef ATOP
+    return runnable;
+}
+
+static void emit_settings_failure(TaskImpl *pimpl, const char *reason) {
+    emit(pimpl, make_log_event(MK_LOG_ERR, reason));
+    emit(pimpl, make_failure_event(ValueError()));
+}
+
+static void emit_settings_warning(TaskImpl *pimpl, const char *reason) {
+    emit(pimpl, make_log_event(MK_LOG_WARNING, reason));
+}
+
+static bool validate_known_settings_shallow(
+        TaskImpl *pimpl, const nlohmann::json &settings) {
+    auto rv = true;
+
+#define VALIDATE(name, type, mandatory)                                        \
```

Comment:
This is either a good argument for or a good argument against the use of macros.

Which is to say: this macro usage is getting complex, and while I think this code is correct, it is worryingly brittle. There's likely not a good way to work around it here, but please start a long-running process imagining a way to accomplish the same goal without using macros-in-macros.

Again: it's okay here for now. But let's try and figure out a better/simpler/clearer way. Ideally one that does not use macros at all.

Reply:
> But let's try and figure out a better/simpler/clearer way. Ideally one that does not use macros at all.

So, an alternative possibility would be to specify in a JSON/YAML file the structure of the messages and then automatically generate code to process them. I have already done some private experiments where I was using YAML, Python, and jinja2 to implement this.

Reply:
@pboothe does the above proposal (Python and jinja2) sounds an improvement over X-macros to you?

Reply:
So, I just had a video chat with @pboothe where he suggests to avoid code generation and instead see whether I can refactor the code to use a factory approach to validate incoming and outgoing messages. The main issue seems to be combining X-macros with long macros with side effects. Perhaps we can combine X-macros with a factory-like code, so that we get more precise type checking and the overall code is more robust to refactoring.